bin/add_dois.py

```diff
@@ -0,0 +1,135 @@
+#! /usr/bin/env python3
+# -*- coding: utf-8 -*-
+#
+# Copyright 2019 Matt Post <post@cs.jhu.edu>
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Used to add ingested DOIs into the Anthology.  
+Does not actually assign DOIs (separate script to manufacture XML to submit to Crossref)
+
+Usage:
+
+  add_dois.py VOLUME_ID
+
+- The ACL Volume ID to add DOIs to (e.g., P17-1, W18-02)
+
+Modifies the XML.  Warns if DOIs already present.  Use -f to force
+"""
+
+import argparse
+import os
+import shutil
+import ssl
+import sys
+import tempfile
+
+DOI_PREFIX = "10.18653/v1/"
+
+from anthology.utils import build_anthology_id, deconstruct_anthology_volume, stringify_children
+
+import lxml.etree as ET
+import urllib.request
+
+def main(args):
+
+    collection_id, volume_id = deconstruct_anthology_volume(args.anthology_volume)
+
+    print(f'Attempting to add DOIs for {args.anthology_volume}', file=sys.stderr)
+
+    # Update XML
+    xml_file = os.path.join(os.path.dirname(sys.argv[0]), '..', 'data', 'xml', f'{collection_id}.xml')
+    tree = ET.parse(xml_file)
+    # add newline to end-of-file if not present
+    if not tree.getroot().tail: tree.getroot().tail = '\n'
+    volume_sequence = str(int(volume_id))
+    volume = tree.getroot().find(f"./volume[@id='{volume_sequence}']")
+    if volume is not None:
+        n = 0
+        volume_title = volume.find(f"./meta/booktitle")
+        print (stringify_children(volume_title))
+        print(f'Identified as {volume_title}', file=sys.stderr)
+
+        # Process frontmatter
+        frontmatter = volume.find('frontmatter')
+        has_doi = False
+        old_doi_text = ""
+
+        for doi in frontmatter.findall('doi'):
+            has_doi = True
+            old_doi_text = doi.text
+
+        if (has_doi == False or args.force == True): # need to assign DOI
```

Comment:
Nitpicking about style: this should just read `if args.force or not has_doi`. Similar with other `== True`/`== False` comparisons.

Reply:
Ok, done in a4f3691